### DISCUSSION TITLE
3.0.0-alpha.2

### DISCUSSION THREAD

**Discussion by mickael-menu** [readium/kotlin-toolkit#481] [Announcements]
## Resources

* [Migration guide](docs/migration-guide.md#300-alpha2)
* [API Reference](https://readium.org/kotlin-toolkit/3.0.0-alpha.2/)

## Changelog

:warning: Please consult [the migration guide](docs/migration-guide.md#300-alpha2) to assist you in handling the breaking changes in this latest major release.

#### Navigator

* The new `HyperlinkNavigator.Listener.shouldFollowInternalLink(Link, LinkContext?)` allows you to handle footnotes according to your preference.
    * By default, the navigator now moves to the footnote content instead of displaying a pop-up as it did in version 2.x.

#### LCP

* You can use `LcpService.injectLicenseDocument()` to insert an LCPL into a package, if you downloaded it manually instead of using `LcpService.acquirePublication()`.

#### Shared

* The `DownloadManager` introduced in version 3.0.0-alpha.1 has been removed due to the Android Download Manager introducing unnecessary complexities in the toolkit. Instead, we chose to enable apps to manually handle an LCP download with `LcpService.injectLicenseDocument()`.

## Commits
* Fix Maven Central publishing and refactor Gradle configurations by @mickael-menu in https://github.com/readium/kotlin-toolkit/pull/452
* Use canonical EPUB HREFs by @mickael-menu in https://github.com/readium/kotlin-toolkit/pull/453
* Extract footnote popups from the Epub navigator by @qnga in https://github.com/readium/kotlin-toolkit/pull/448
* Implement sniffing of CSS and JS files by @io7m in https://github.com/readium/kotlin-toolkit/pull/455
* Remove `android:clearTaskOnLaunch` by @mickael-menu in https://github.com/readium/kotlin-toolkit/pull/459
* Cleanup Gradle and Version files, fix Dokka by @stevenzeck in https://github.com/readium/kotlin-toolkit/pull/456
* Use Flow instead of LiveData by @stevenzeck in https://github.com/readium/kotlin-toolkit/pull/461
* Bump `css-selector-generator` script dependency by @mickael-menu in https://github.com/readium/kotlin-toolkit/pull/466
* Remove the download manager by @qnga in https://github.com/readium/kotlin-toolkit/pull/465
* Small fix by @qnga in https://github.com/readium/kotlin-toolkit/pull/467
* Use material3 for Compose views by @stevenzeck in https://github.com/readium/kotlin-toolkit/pull/460
* Add LcpService.injectLicenseDocument by @qnga in https://github.com/readium/kotlin-toolkit/pull/473
* Fix EPUB deobfuscation by @mickael-menu in https://github.com/readium/kotlin-toolkit/pull/475
* Prevent infinite recursion in format sniffing by @qnga in https://github.com/readium/kotlin-toolkit/pull/476
* Fix TTS player crash by @qnga in https://github.com/readium/kotlin-toolkit/pull/477
* Update media3 and fix some issues with MediaService by @qnga in https://github.com/readium/kotlin-toolkit/pull/478
* Simplify LcpService.injectLicense by @qnga in https://github.com/readium/kotlin-toolkit/pull/479
* 3.0.0-alpha.2 by @mickael-menu in https://github.com/readium/kotlin-toolkit/pull/480

## New Contributors
* @io7m made their first contribution in https://github.com/readium/kotlin-toolkit/pull/455

**Full Changelog**: https://github.com/readium/kotlin-toolkit/compare/3.0.0-alpha.1...3.0.0-alpha.2

<hr /><em>This discussion was created from the release <a href='https://github.com/readium/kotlin-toolkit/releases/tag/3.0.0-alpha.2'>3.0.0-alpha.2</a>.</em>

**Comment by radusalagean** [comment]
@mickael-menu Back in `v2.4.0` I used to override `onResourceLoadFailed` from `PdfNavigatorFragment.Listener` in order to track errors like OutOfMemory. Is there an alternative to this for v3? Thanks!
![Screenshot 2024-04-15 at 19 28 20](https://github.com/readium/kotlin-toolkit/assets/11408459/45387898-34ae-4338-a152-49a0849fe6a6)

**Reply by mickael-menu** [reply]
Yes, there's a new `onResourceLoadFailed` in `Navigator.Listener`. The error is slightly different though:

https://github.com/readium/kotlin-toolkit/blob/c9a09acc629713eba2fd83878464fe126407346a/readium/navigator/src/main/java/org/readium/r2/navigator/Navigator.kt#L53-L56

```kotlin
override fun onResourceLoadFailed(href: Url, error: ReadError) {
    val message = when (error) {
        is ReadError.OutOfMemory -> ...
    }
}
```
